pennylane/io/qualtran_io.py

```diff
@@ -0,0 +1,285 @@
+
+    temp_register_dict = {reg.name: reg.total_bits() for reg in registers}
+
+    return qml.registers(temp_register_dict)
+
+
+class FromBloq(Operation):
+    r"""
+    An adapter for using a `Qualtran bloq <https://qualtran.readthedocs.io/en/latest/bloqs/index.html#bloqs-library>`_
+    as a PennyLane :class:`~.Operation`.
+
+    Args:
+        bloq (qualtran.Bloq): an initialized Qualtran bloq to be wrapped as a PennyLane operator
+        wires (WiresLike): The wires the operator acts on. This can be determined by using the
+            signature of the Bloq as ``bloq.signature.n_qubits()``.
+
+    Raises:
+        TypeError: bloq must be an instance of :code:`~.Bloq`.
+
+    **Example**
+
+    Given a qualtran bloq:
+
+    >>> from qualtran.bloqs.basic_gates import CNOT
+
+    >>> qualtran_cnot = qml.FromBloq(CNOT(), wires=[0, 1])
+    >>> qualtran_cnot.matrix()
+    array([[1.+0.j, 0.+0.j, 0.+0.j, 0.+0.j],
+       [0.+0.j, 1.+0.j, 0.+0.j, 0.+0.j],
+       [0.+0.j, 0.+0.j, 0.+0.j, 1.+0.j],
+       [0.+0.j, 0.+0.j, 1.+0.j, 0.+0.j]])
+
+    A simple example showcasing how to use ``qml.FromBloq`` inside a device:
+
+    .. code-block::
+
+        from qualtran.bloqs.phase_estimation import RectangularWindowState, TextbookQPE
+        from qualtran.bloqs.chemistry.trotter.ising import IsingXUnitary, IsingZZUnitary
+        from qualtran.bloqs.chemistry.trotter.trotterized_unitary import TrotterizedUnitary
+
+        # Parameters for the TrotterizedUnitary
+        nsites = 5
+        j_zz = 2
+        gamma_x = 0.1
+        zz_bloq = IsingZZUnitary(nsites=nsites, angle=0.02 * j_zz)
+        x_bloq = IsingXUnitary(nsites=nsites, angle=0.01 * gamma_x)
+        trott_unitary = TrotterizedUnitary(
+            bloqs=(x_bloq, zz_bloq), indices=(0, 1, 0), coeffs=(0.5 * gamma_x, j_zz, 0.5 * gamma_x), timestep=0.01
+        )
+
+        # Instantiate the TextbookQPE and pass in our unitary
+        textbook_qpe = TextbookQPE(trott_unitary, RectangularWindowState(3))
+
+        # Execute on device
+        dev = qml.device("default.qubit")
+        @qml.qnode(dev)
+        def circuit():
+            qml.FromBloq(textbook_qpe, wires=range(textbook_qpe.signature.n_qubits()))
+            return qml.probs(wires=[5, 6, 7])
+
+    >>> circuit()
+    array([0.94855734, 0.01291602, 0.00431044, 0.00267219, 0.00237645,
+    0.00290337, 0.00526717, 0.02099701])
+
+    .. details::
+        :title: Usage Details
+
+        The decomposition of a bloq wrapped in ``qml.FromBloq`` may use more wires than expected.
+        For example, when we wrap Qualtran's ``CZPowGate``, we get
+
+        >>> from qualtran.bloqs.basic_gates import CZPowGate
+        >>> qml.FromBloq(CZPowGate(0.468, eps=1e-11), wires=[0, 1]).decomposition()
+        [FromBloq(And, wires=Wires([0, 1, 'alloc_free2'])),
+        FromBloq(Z**0.468, wires=Wires(['alloc_free2'])),
+        FromBloq(And†, wires=Wires([0, 1, 'alloc_free2']))]
```

Comment:
Minor - where does `alloc_free1` go in this case?

Reply:
The indexing for alloc_free starts at the length of the total_wires. So if the signature suggests 2 total wires, alloc_free starts at 2 and goes onwards. Otherwise I have to create a whole separate "count" outside of the scope to track the number of these type of wires, which seems a bit excessive.

Reply:
Might want to mention it here then.